wire-ios-sync-engine/Source/Use cases/ImportBackupUseCase/ImportBackupUseCase.swift

```diff
@@ -0,0 +1,189 @@
+//
+// Wire
+// Copyright (C) 2025 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireCrypto
+import WireLogging
+import WireSystem
+import ZipArchive
+
+struct ImportBackupUseCase: ImportBackupUseCaseProtocol {
+
+    let userSession: () -> ZMUserSession?
+    let dispatchGroup: ZMSDispatchGroup
+    let fileArchiver: ImportBackupFileArchiverProtocol
+    let entityStorage: ImportBackupEntityStorageProtocol
+    let appStateUpdater: ImportBackupAppStateUpdaterProtocol
+
+    let sharedContainerURL: URL
+    let logger: WireLogger
+
+    private let workerQueue = DispatchQueue(label: "import-backup")
+
+    func invoke(url: URL, password: String) async throws {
+
+        switch BackupFileExtensions(rawValue: url.pathExtension) {
+
+        case .fileExtensionWithUnderscore, .fileExtensionWithHyphen:
+            try await importIOSBackup(url, password)
+
+        case nil:
+            throw BackupError.invalidFileExtension
+        }
+    }
+
+    private func importIOSBackup(_ url: URL, _ password: String) async throws {
+
+        let selfClientBackup: [String: Any]
+        let account: Account
+        if let userSession = userSession() {
+
+            account = userSession.account
+
+            appStateUpdater.reportImportProgress(progress: 0.5)
+
+            let unzippedURL = try await decryptAndUnzipBackup(
+                url: url,
+                password: password,
+                accountID: userSession.account.userIdentifier
+            )
+
+            selfClientBackup = await userSession.managedObjectContext.perform {
+                userSession.selfUserClient?.backup() ?? [:]
+            }
+
+            await appStateUpdater.reportMigrationNeeded()
+
+            // user session will be destroyed now
+            _ = try await entityStorage.replacePersistentStore(
+                accountIdentifier: userSession.account.userIdentifier,
+                from: unzippedURL,
+                applicationContainer: sharedContainerURL,
+                dispatchGroup: dispatchGroup
+            )
+
+        } else {
+            throw BackupError.noActiveAccount
+        }
```

Comment:
nit-pic: guard else the to handle the error first

Reply:
I did it like this so that we don't hold a the strong to `userSession` for the whole use case.
But I will try to refactor it.